File: reader/csharp/I18N.cs
using ProtoBuf;
using System;
using System.Collections.Generic;
using System.IO;

[Serializable]
[ProtoContract]
public class Language : PBDataModel
{
    [ProtoMember(1)]
    public string Id { get; set; }

    [ProtoMember(2)]
    public string Text { get; set; }
}

[Serializable]
[ProtoContract]
public class Language_ARRAY : PBDataModels
{
    [ProtoMember(1)]
    public List<Language> Items { get; set; }
}

public class I18N : DataContainer<string, Language> 
{
    public static string CurrentLanguage { get; private set; }
    public static void SetLanguage(string lan = "cn")
    {
        if (CurrentLanguage != "")
        {
            Instance.Clear();
        }

        CurrentLanguage = lan;

        Instance.Initial(null, GetFilename);
    }

    protected static string GetFilename(string typeName)
    {
        var datafile = Path.Combine(DataAccess.DataDir, typeName.ToLower());
        return $"{datafile}.{CurrentLanguage}";
    }
    
    public static string Translate(int key, params object[] args)
    {
        return Translate(key.ToString(), args);
    }

    public static string Translate(string key, params object[] args)
    {
        Language lan;
        if (Instance.ItemMap.TryGetValue(key, out lan))
        {
            if (args == null || args.Length == 0)
            {
                return lan.Text;
            }
            else
            {
                return string.Format(lan.Text, args);
            }
        }

        return null;
    }
}

File: reader/typescript/I18N.ts
import { DataAccess, DataTable, IDataArray, IMessage } from "./DataAccess";
import { Language, Language_ARRAY } from "./Lang";

export class I18N extends DataTable<Language>
{
    private static _inst;
    static get inst(): I18N {
        if(!this._inst) {
            this._inst = new I18N();
        }
        return this._inst;
    }

    private constructor(keyName = "ID") {
        super(Language, keyName);

        this.initial(this.dataType, null, I18N.getFilename);
    }

    protected load() : Language[] {
        var buffer = this.onLoadData("Language_ARRAY");
        var message = Language_ARRAY.decode(buffer);
        return (message as IDataArray).Items;        
    }

    public static currentLanguage: string = "cn";
    public static setLanguage(lan = "cn")
    {
        if (this.currentLanguage != "")
        {
            this.inst.clear();
        }

        this.currentLanguage = lan;
    }

    protected static getFilename(typeName: string): string
    {
        typeName = typeName.replace("_ARRAY", "");
        var datafile = DataAccess.dataDir + typeName.toLocaleLowerCase();
        return `${datafile}.${I18N.currentLanguage}`;
    }

    public static translate(key: string|number): string
    {
        let lan:Language = this.inst.itemMap[key];
        if (lan)
        {
            return lan.Text;
        }

        return null;
    }
}
